python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -613,15 +613,34 @@ def evaluate_for_asset(
         """
         from .asset_condition import AssetConditionResult
 
-        handled_subset = self.get_handled_subset(context)
-        unhandled_candidates = (
-            context.candidate_subset
-            & handled_subset.inverse(
-                context.partitions_def, context.evaluation_time, context.instance_queryer
+        if context.asset_key in context.asset_graph.root_materializable_or_observable_asset_keys:
+            handled_subset = self.get_handled_subset(context)
+            unhandled_candidates = (
+                context.candidate_subset
+                & handled_subset.inverse(
+                    context.partitions_def, context.evaluation_time, context.instance_queryer
+                )
+                if handled_subset.size > 0
+                else context.candidate_subset
```

Comment:
netting `context.candidate_subset & context.candidate_subset` in the `else` case feels like cleverness just for the sake of compacting this to a single expression

Reply:
yeah good point -- I'll merge as is because this gets updated somewhat in a PR that's stacked on top of this (but can be further simplified there: https://github.com/dagster-io/dagster/pull/19483/files#diff-d51d8c2de97cbb7069cdebdeec5939c85254e33628bc74f1cc7ea6f03b0ccc04)